File: python/81_path_sum_two_ways.py
def path_sum_two_ways(matrix: list) -> int:
    r = len(matrix)
    c = len(matrix[0])
    
    for row in range(0, r):
        for col in range(0, c):
            if row == 0 and col == 0: # starting cell
                continue
            elif row == 0: # can only reach cell from left neighbor
                matrix[row][col] += matrix[row][col - 1]
            elif col == 0: # can only reach cell from above neighbor
                matrix[row][col] += matrix[row - 1][col]
            else:
                matrix[row][col] += min(matrix[row][col - 1], matrix[row - 1][col])
            
    return matrix[r - 1][c - 1]
    

if __name__ == "__main__":
    f = open("81_matrix.txt", 'r')
    matrix = []
    for line in f.readlines():
        matrix.append([int(num) for num in line.strip().split(",")])
    f.close()
    
    # matrix = [
        # [131, 673, 234, 103, 18],
        # [201, 96, 342, 965, 150],
        # [630, 803, 746, 422, 111],
        # [537, 699, 497, 121, 956],
        # [805, 732, 524, 37, 331]
    # ]
    
    print(path_sum_two_ways(matrix))

File: cpp/81_path_sum_two_ways.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <cstring>

int path_sum_two_ways(std::vector<std::vector<int>> matrix){
    for (int r = matrix.size() - 1; r >= 0; r--){
        for (int c = matrix[0].size() - 1; c >= 0; c--){
            if (r == matrix.size() - 1 and c == matrix[0].size() - 1) continue;
            // only one subsequent cell possible for cells in bottom row or rightmost column
            if (r == matrix.size() - 1){
                matrix[r][c] += matrix[r][c + 1];
            }
            else if (c == matrix[0].size() - 1){
                matrix[r][c] += matrix[r + 1][c];
            }
            // add minimum of cells immediately right and below current cell
            else {
                matrix[r][c] += std::min(matrix[r + 1][c], matrix[r][c + 1]);
            }
        }
    }
    // top left cell contains minimal path sum
    return matrix[0][0];
}

int main(){
    std::ifstream file("/uploads/81_matrix.txt");
    std::string s;
    std::vector<std::vector<int>> matrix;
    
    // tokenize each line into vector of ints
    while (std::getline(file, s)){
        std::vector<int> row;
        char *pch = std::strtok(&s[0], ",");
        while (pch != NULL){
            std::string num(pch);
            row.push_back(stoi(num));
            pch = std::strtok(NULL, ",");
        }
        matrix.push_back(row);
    }
    file.close();
    
    std::cout << path_sum_two_ways(matrix) << "\n";
}
